### DISCUSSION TITLE
Performance for large database

### DISCUSSION THREAD

**Discussion by fbraennstroem** [papis/papis#805] [General]
Hi,
what is your experience of the performance of papis with large database? Not sure, where a large database really starts, so maybe there are some users which have on in the range of 1000, 5000, 10000 or more pdfs and can say something about it, when searching for a pdf, editing any yaml file or adding new documents!?
Thank you!
Fabian

**Comment by alexfikl** [comment]
It's not fast-fast, but it should be pretty good from my experience.

For a data point, here's how it looks on my system. I'm running this on a Dell Latitude 5530 (12th Gen Intel i7, 16GB RAM, 512GB SSD) on Linux (Python 3.11, `papis` database backend). This repeatedly runs a command that lists about 200 documents from a library of about 2500 documents
```
Benchmark 1: papis -l papers list -a fractional
Time (mean ± σ):     300.2 ms ±   3.0 ms    [User: 245.5 ms, System: 53.9 ms]
Range (min … max):   296.1 ms … 306.7 ms    10 runs
```
and about 20 documents from a library of about 700 documents
```
Benchmark 1: papis -l books list -a fractional
Time (mean ± σ):     211.5 ms ±   3.1 ms    [User: 177.3 ms, System: 33.4 ms]
Range (min … max):   207.4 ms … 218.9 ms    14 runs
```

From what I remember benchmarking, it actually spends most startup time just loading all the various plugins. We're using `stevedore` for that, which does some caching and stuff, but it's still not instant.

Scaling up to 10k documents or more shouldn't be too bad either. Generally a command does:
* A query through the database: this could likely be optimized and parallelized better.
* Some action on the resulting documents (e.g. export to BibTeX). Queries generally shouldn't match thousands of documents, so this shouldn't be a bottleneck.

**Comment by fbraennstroem** [comment]
Thank you for your quick reply!
Interesting, I was creating some dummy directory with yaml and pdf files, but it seems to always to re-index the cache with a papis database.
Switching to woosh, does not do anything, I mean, then not any document is listed.
And trying to run `papis cache clear`, it seems that this (no `papis cache` command exists) does not exist... hm, probably I did something weird with my installation.
<img width="425" alt="image" src="https://github.com/papis/papis/assets/83709063/fecda1b6-5007-4083-adc3-3373a5a87b59">

I figured out that I was still using the old 0.11 version due to a path issue in my installation. But now, running with 0.13, any `papis cache` is not available as well.

I will look into my installation first...

Thank you!

**Reply by alexfikl** [reply]
The `papis cache` command is only available in the unreleased version (which will be 0.14). You can use that, but you'll have to install from git. In older versions, 
```
papis --clear-cache
```
should do what you want. I'm not sure why it would keep re-indexing the library though. Did you add that library to your configuration file or just pointed papis at a directory?
